plugins/modules/source_control/gitlab/gitlab_service.py

```diff
@@ -0,0 +1,485 @@
+        # python2.6 remove-me:
+        filtered_params = {}
+        for k, v in iteritems(params):
+            if k not in self.credentialParams:
+                filtered_params[k] = v
+
+        if prev_attr['active'] != active:
+            self._module.debug("active status differs")
+            return False
+        if self.has_credentials(params):
+            self._module.debug("has credentials: assuming difference")
+            return False
+        if not self.events_equal(prev_attr, self.expand_events(events)):
+            self._module.debug("events differs: %s != %s" % (prev_attr, events))
+            return False
+        if prev_attr['properties'] != filtered_params:
+            self._module.debug("services attributes differs: %s != %s" % (prev_attr['properties'], filtered_params))
+            return False
+        return True
+
+    def setattrs(self, obj, new_attr):
+        for k, v in new_attr.items():
+            setattr(obj, k, v)
+
+
+def main():
+    definitions = init_definitions()
+
+    base_spec = basic_auth_argument_spec()
+    base_spec.update(dict(
+        api_token=dict(type='str', no_log=True),
+        project=dict(required=True),
+        service=dict(required=True, type='str', choices=list(definitions.keys())),
+        active=dict(required=False, default=True, type='bool'),
+        params=dict(required=False, type='dict'),
+        events=dict(required=False, type='list', elements='str', default=GitLab_Services.HOOK_EVENTS, choices=GitLab_Services.HOOK_EVENTS),
+        state=dict(default='present', choices=['present', 'absent']),
+    ))
+
+    constraints = dict(
+        mutually_exclusive=[
+            ['api_username', 'api_token'],
+            ['api_password', 'api_token']
+        ],
+        required_together=[
+            ['api_username', 'api_password']
+        ],
+        required_one_of=[
+            ['api_username', 'api_token']
+        ],
+        required_if=[
+            ['state', 'present', ['params']]
+        ],
+        supports_check_mode=True
+    )
+
+    stub_init = AnsibleModule(argument_spec=base_spec, **constraints)
+    service = stub_init.params['service']
+    state = stub_init.params['state']
+    if state == 'present':
+        # since we know the service (which has been validated), recreate a module_specs to validate suboptions
+        # sub_arg_specs = {i: definitions[service][i] for i in definitions[service] if i != '_events'}
+        # python2.6 remove-me:
+        sub_arg_specs = {}
+        for i in iteritems(definitions[service]):
+            if i != '_events':
+                sub_arg_specs[i] = definitions[service][i]
+
+        base_spec['params'] = dict(required=False, type='dict', options=sub_arg_specs)
+        if '_events' in definitions[service]:
+            base_spec['events'] = dict(required=True, type='list', elements='str', choices=definitions[service]['_events'])
+        module = AnsibleModule(argument_spec=base_spec, **constraints)
```

Comment:
I don't think `AnsibleModule` is meant to be used like this.

Reply:
You mean re-instanciating an `AnsibleModule` a second time in order to further validate parameters based on a first pass (because a complex `dict` parameter depends upon the value of another)... Then yes, I admit it's indeed exotic (and its the first plugin to use such a construct).
That being said we already discussed the question of parameters (and the benefits of validation) (and so far I couldn't think of any negative side effect for the double `AnsibleModule` instanciation)

Reply:
I really don't think it is a good idea to do this. I also asked on #ansible-devel, and one of the core team members confirmed that this is something that they support. It's basically undefined behavior, and use at your own risk.

Reply:
> I really don't think it is a good idea to do this. I also asked on #ansible-devel, and one of the core team members confirmed that this is something that they **not** support. It's basically undefined behavior, and use at your own risk.

Damn, I missed the most important word in this: this is something they do **not** support. While it works now, it can stop working at a random point in the future.

Reply:
Don't you think it's very convenient for the user and that if it were to stop working, we could adapt it quite easily?
The usability benefit are huge (and I think this could be useful to other module if it started to be supported in the future).
Shouldn't we try it? And if not, what would you suggest instead?

Reply:
> Don't you think it's very convenient for the user

It definitely is! But that can (and should) also be achieved differently than relying on undefined behavior.

> and that if it were to stop working, we could adapt it quite easily?

Well, it depends on who will do the work. Such cleanup work often falls back to collection maintainers because the people who originally implemented it are no longer around to clean it up.

> The usability benefit are huge (and I think this could be useful to other module if it started to be supported in the future).

The benefits of extra validation are indeed great, but doing that by instantiating AnsibleModule twice is IMO the wrong solution for that problem.

> And if not, what would you suggest instead?

Implement the validation manually, preferably as a module_utils so it can be used by other modules as well. Useful parts for that already exist in Ansible(-base/-core)'s module_utils, but unfortunately they change a lot between the Ansible versions (2.9, 2.10, 2.11) and thus most of them cannot be used. What can be used are the things in https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/common/validation.py that were already present in https://github.com/ansible/ansible/blob/stable-2.9/lib/ansible/module_utils/common/validation.py, which covers all `check_type_xxx` functions, and `check_required_arguments`.